README.md
The documentation is available here: https://help.blueshift.com/hc/en-us/articles/115002731514-Mobile-SDK
test 1


release_notes
base64 encoded content 

BlueShift-iOS-SDK/BlueShift.h
//
//  BlueShift.h
//  BlueShift
//
//  Copyright (c) Blueshift. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "BlueShiftConfig.h"
#import "BlueShiftDeviceData.h"
#import "BlueShiftAppDelegate.h"
#import "BlueShiftPushDelegate.h"
#import "BlueShiftDeepLink.h"
#import "BlueShiftPushParamDelegate.h"
#import "BlueShiftNetworkReachabilityManager.h"
#import "BlueShiftSubscriptionState.h"
#import "BlueShiftRequestOperation.h"
#import "BlueShiftRequestQueue.h"
#import "BlueShiftRoutes.h"
#import "BlueShiftUserInfo.h"
#import "BlueShiftTrackEvents.h"
#import "BlueShiftProduct.h"
#import "BlueShiftSubscription.h"
#import "BlueShiftHttpRequestBatchUpload.h"
#import "BlueShiftBatchUploadConfig.h"
#import "BlueShiftAppData.h"
#import "SDKVersion.h"
#import "BlueShiftUserNotificationSettings.h"
#import "BlueShiftUserNotificationCenterDelegate.h"
#import "BlueshiftEventAnalyticsHelper.h"
#import "BlueshiftInAppNotificationRequest.h"
#import "BlueShiftLiveContent.h"

@class BlueShiftDeviceData;
@class BlueShiftAppDelegate;
@class BlueShiftUserNotificationCenterDelegate;
@class BlueShiftUserInfo;
@class BlueShiftConfig;
@interface BlueShift : NSObject

@property (nonatomic, strong) BlueShiftConfig *config;
@property BlueShiftDeviceData *deviceData;
@property BlueShiftAppData *appData;
@property (nonatomic, strong) BlueShiftUserInfo *userInfo;
@property (nonatomic, strong) BlueShiftUserNotificationSettings *userNotification;
@property (nonatomic, strong, readwrite) NSTimer *inAppMsgAPITimer;
@property (nonatomic, strong, readwrite) NSTimer *inAppDBTimer;
@property NSString *deviceToken;

+ (instancetype)sharedInstance;
+ (void) initWithConfiguration:(BlueShiftConfig *)config;
+ (void) autoIntegration;
- (void) setPushDelegate: (id) obj;
- (void) setPushParamDelegate: (id) obj;
- (NSString *) getDeviceToken;
- (void) setDeviceToken;
- (void) createInAppNotification:(NSDictionary *)dictionary forApplicationState:(UIApplicationState)applicationState;
- (void)registerForInAppMessage:(NSString *)displayPage;
- (void)unregisterForInAppMessage;

@property BlueShiftAppDelegate *appDelegate;
@property BlueShiftUserNotificationCenterDelegate *userNotificationDelegate;



// track events functions ...
- (void)identifyUserWithDetails:(NSDictionary *)details canBatchThisEvent:(BOOL)isBatchEvent;

- (void)identifyUserWithEmail:(NSString *)email andDetails:(NSDictionary *)details canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackScreenViewedForViewController:(UIViewController *)viewController canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackScreenViewedForViewController:(UIViewController *)viewController withParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackProductViewedWithSKU:(NSString *)sku andCategoryID:(NSInteger)categoryID canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackProductViewedWithSKU:(NSString *)sku andCategoryID:(NSInteger)categoryID withParameter:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackAddToCartWithSKU:(NSString *)sku andQuantity:(NSInteger)quantity canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackAddToCartWithSKU:(NSString *)sku andQuantity:(NSInteger)quantity andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackCheckOutCartWithProducts:(NSArray *)products andRevenue:(float)revenue andDiscount:(float)discount andCoupon:(NSString *)coupon canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackCheckOutCartWithProducts:(NSArray *)products andRevenue:(float)revenue andDiscount:(float)discount andCoupon:(NSString *)coupon andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackProductsPurchased:(NSArray *)products withOrderID:(NSString *)orderID andRevenue:(float)revenue andShippingCost:(float)shippingCost andDiscount:(float)discount andCoupon:(NSString *)coupon canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackProductsPurchased:(NSArray *)products withOrderID:(NSString *)orderID andRevenue:(float)revenue andShippingCost:(float)shippingCost andDiscount:(float)discount andCoupon:(NSString *)coupon andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackPurchaseCancelForOrderID:(NSString *)orderID canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackPurchaseCancelForOrderID:(NSString *)orderID andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackPurchaseReturnForOrderID:(NSString *)orderID andProducts:(NSArray *)products canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackPurchaseReturnForOrderID:(NSString *)orderID andProducts:(NSArray *)products andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackProductSearchWithSkuArray:(NSArray *)skuArray andNumberOfResults:(NSInteger)numberOfResults andPageNumber:(NSInteger)pageNumber andQuery:(NSString *)query andFilters:(NSDictionary *)filters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackProductSearchWithSkuArray:(NSArray *)skuArray andNumberOfResults:(NSInteger)numberOfResults andPageNumber:(NSInteger)pageNumber andQuery:(NSString *)query andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackProductSearchWithSkuArray:(NSArray *)skuArray andNumberOfResults:(NSInteger)numberOfResults andPageNumber:(NSInteger)pageNumber andQuery:(NSString *)query andFilters:(NSDictionary *)filters andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackEmailListSubscriptionForEmail:(NSString *)email canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackEmailListSubscriptionForEmail:(NSString *)email andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackEmailListUnsubscriptionForEmail:(NSString *)email canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackEmailListUnsubscriptionForEmail:(NSString *)email andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionInitializationForSubscriptionState:(BlueShiftSubscriptionState)subscriptionState andCycleType:(NSString *)cycleType andCycleLength:(NSInteger)cycleLength andSubscriptionType:(NSString *)subscriptionType andPrice:(float)price andStartDate:(NSTimeInterval)startDate canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionInitializationForSubscriptionState:(BlueShiftSubscriptionState)subscriptionState andCycleType:(NSString *)cycleType andCycleLength:(NSInteger)cycleLength andSubscriptionType:(NSString *)subscriptionType andPrice:(float)price andStartDate:(NSTimeInterval)startDate andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionPauseWithBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionPauseWithParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionUnpauseWithBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionUnpauseWithParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionCancelWithBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackSubscriptionCancelWithParamters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackEventForEventName:(NSString *)eventName canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackEventForEventName:(NSString *)eventName andParameters:(NSDictionary *)parameters canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackPushClickedWithParameters:(NSDictionary *)userInfo canBatchThisEvent:(BOOL)isBatchEvent;

- (void)trackPushViewedWithParameters:(NSDictionary *)userInfo canBacthThisEvent:(BOOL)isBatchEvent;

- (void)trackInAppNotificationDeliveredWithParameter:(NSDictionary *)notification canBacthThisEvent:(BOOL)isBatchEvent;

- (void)trackInAppNotificationShowingWithParameter:(NSDictionary *)notification canBacthThisEvent:(BOOL)isBatchEvent;

- (void)trackInAppNotificationButtonTappedWithParameter:(NSDictionary *)notification canBacthThisEvent:(BOOL)isBatchEvent;

- (void)trackInAppNotificationDismissWithParameter:(NSDictionary *)notificationPayload
                                 canBacthThisEvent:(BOOL)isBatchEvent;
- (void)displayInAppNotification;

- (void)fetchInAppNotificationFromAPI:(void (^)(void))handler;

@end


BlueShift-iOS-SDK/InApps/BlueShiftInAppNotificationManager.m
//
//  BlueShiftInAppNotificationManager.m
//  BlueShift-iOS-Extension-SDK
//
//  Created by shahas kp on 10/07/19.
//
#import <StoreKit/StoreKit.h>
#import "BlueShiftInAppNotificationManager.h"
#import "ViewControllers/Templates/BlueShiftNotificationWebViewController.h"
#import "ViewControllers/Templates/BlueShiftNotificationModalViewController.h"
#import "ViewControllers/Templates/BlueShiftNotificationSlideBannerViewController.h"
#import "Models/InAppNotificationEntity.h"
#import "BlueShiftAppDelegate.h"
#import "BlueShiftInAppTriggerMode.h"
#import "BlueShiftInAppNotificationConstant.h"

#define THRESHOLD_FOR_UPCOMING_IAM  (30*60)         // 30 min set for time-being.


@interface BlueShiftInAppNotificationManager() <BlueShiftNotificationDelegate>

/* In-App message timer for handlin upcoming messages */
@property (nonatomic, strong, readwrite) NSTimer *inAppMsgTimer;

/* private object context */
@property (nonatomic, strong, readwrite) NSManagedObjectContext *privateObjectContext;

@end

@implementation BlueShiftInAppNotificationManager

// init
- (void)load {
    self.notificationControllerQueue = [NSMutableArray new];
    
    /* create timer for upcoming events */
    [self startInAppMessageLoadTimer];
    
    /* register for app background / foreground notification */
    
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(OnApplicationEnteringBackground:)
                                                 name:UIApplicationDidEnterBackgroundNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(OnApplicationEnteringForeground:)
                                                 name:UIApplicationWillEnterForegroundNotification
                                               object:nil];
    
    
}

- (void) OnApplicationEnteringBackground:(NSNotification *)notification {
    /* stop the timer once app enters background */
    
    [self stopInAppMessageLoadTimer];
}

- (void) OnApplicationEnteringForeground:(NSNotification *)notification {
    /* start the timer once app enters foreground */
    
    [self startInAppMessageLoadTimer];
    
}


- (void) addInAppNotificationToDataStore: (NSDictionary*)payload forApplicationState:(UIApplicationState)applicationState {
    
    /* creating a private context */
    if (nil == self.privateObjectContext) {
        self.privateObjectContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
    }
    
    BlueShiftAppDelegate *appDelegate = (BlueShiftAppDelegate *)[BlueShift sharedInstance].appDelegate;
    NSManagedObjectContext *masterContext;
    if (appDelegate) {
        @try {
            masterContext = appDelegate.managedObjectContext;
        }
        @catch (NSException *exception) {
            NSLog(@"Caught exception %@", exception);
        }
    }
    if(masterContext) {
        NSEntityDescription *entity;
        @try {
            entity = [NSEntityDescription entityForName:@"InAppNotificationEntity" inManagedObjectContext:masterContext];
        }
        @catch (NSException *exception) {
            NSLog(@"Caught exception %@", exception);
        }
        if(entity != nil) {
            
            NSManagedObjectContext *context = appDelegate.managedObjectContext;
            
            InAppNotificationEntity *inAppNotificationEntity = [[InAppNotificationEntity alloc] initWithEntity:entity insertIntoManagedObjectContext:context];
            
            if(inAppNotificationEntity != nil) {
                printf("%f NotificationMgr: Inserting the payload \n", [[NSDate date] timeIntervalSince1970]);
                [inAppNotificationEntity insert:payload inContext:self.privateObjectContext handler:^(BOOL status) {
                    if(status) {
                        printf("%f NotificationMgr: Insert Done. Loading from DB \n", [[NSDate date] timeIntervalSince1970]);
                        if (applicationState == UIApplicationStateActive ||
                            applicationState == UIApplicationStateInactive) {
                            [self fetchInAppNotificationsFromDataStore: BlueShiftInAppTriggerNow];
                        } else {
                            NSLog(@"NotificationMgr:: Saving in-app msg just saved in CoreDataApp. AppState = %d" , applicationState);
                        }
                    }
                }];
            }
        }
    }
}


- (void)removeInAppNotificationFromDB:(NSManagedObjectID *) entityItem {
    
    BlueShiftAppDelegate *appDelegate = (BlueShiftAppDelegate *)[BlueShift sharedInstance].appDelegate;
    
    NSManagedObjectContext *masterContext;
    if (appDelegate) {
        @try {
            masterContext = appDelegate.managedObjectContext;
        }
        @catch (NSException *exception) {
            NSLog(@"Caught exception %@", exception);
        }
    }
    
    if (entityItem != nil) {
        
        NSManagedObjectContext *context = self.privateObjectContext;
        context.parentContext = masterContext;
        
        @try {
            if(context && [context isKindOfClass:[NSManagedObjectContext class]]) {
                [context performBlock:^{
                    NSManagedObject* pManagedObject =  [context objectWithID: entityItem];
                    
                    @try {
                        [context deleteObject: pManagedObject];
                    }
                    @catch (NSException *exception) {
                        NSLog(@"Caught exception %@", exception);
                    }
                    [context performBlock:^{
                        NSError *saveError = nil;
                        if(context && [context isKindOfClass:[NSManagedObjectContext class]]) {
                            [context save:&saveError];
                            
                            if(masterContext && [masterContext isKindOfClass:[NSManagedObjectContext class]]) {
                                [masterContext performBlock:^{
                                    NSError *error = nil;
                                    [masterContext save:&error];
                                }];
                            }
                        }
                    }];
                }];
            } else {
            }
        }
        @catch (NSException *exception) {
            NSLog(@"Caught exception %@", exception);
        }
    }
}




- (void)fetchInAppNotificationsFromDataStore: (BlueShiftInAppTriggerMode) triggerMode  {
    [InAppNotificationEntity fetchAll: triggerMode withHandler:^(BOOL status, NSArray *results) {
        
        if (status) {
            NSArray* filteredResults = [self filterInAppNotificationResults:results withTriggerMode:triggerMode];
            
            for(int i = 0; i < [filteredResults count]; i++) {
                InAppNotificationEntity *entity = [filteredResults objectAtIndex:i];
                [self createNotificationFromDictionary: entity];
            }
        }
    }];
}


- (NSArray *) filterInAppNotificationResults: (NSArray*) results withTriggerMode:(BlueShiftInAppTriggerMode) triggerMode {
    
    /* get the current time (since 1970) */
    NSTimeInterval currentTime =  [[NSDate date] timeIntervalSince1970];
    
    NSArray *outResults = nil;
    
    if (BlueShiftInAppTriggerUpComing == triggerMode)
    {
        NSMutableArray* filteredResults = [[NSMutableArray alloc] init];
        
        for(int i = 0; i < [results count]; i++) {
            InAppNotificationEntity *entity = [results objectAtIndex:i];
            
            
            double endTime = [entity.endTime doubleValue];
            double startTime = [entity.startTime doubleValue];
            
            if (currentTime - THRESHOLD_FOR_UPCOMING_IAM > endTime) {
                /* discard notification if its expired. */
    
                //TODO: remove from db
                
                printf("\nUpcoming:: Discarded Current time of IAM = %f endTime = %f", currentTime, endTime);
                
            } else if (startTime > currentTime) {
                /* Wait for (startTime-currentTime) before IAM is shown */
                
                printf("\nUpcoming:: Wait further Current time of IAM = %f startTime = %f", currentTime, startTime);
                
            } else {
                
                printf("\nUpcoming:: Display Current time of IAM = %f startTime = %f", currentTime, startTime);
                
                [filteredResults addObject:entity];
            }
        }
        outResults = [NSArray arrayWithArray: filteredResults];
    } else {
        outResults = results;
    }
    return outResults;
}



// Method to start In-App message loading timer
- (void)startInAppMessageLoadTimer {
    if (nil == self.inAppMsgTimer) {
        self.inAppMsgTimer = [NSTimer scheduledTimerWithTimeInterval:10
                                                          target:self
                                                        selector:@selector(handlePendingInAppMessage)
                                                        userInfo:nil
                                                         repeats:YES];
    }
}

// Method to stop In-App message loading timer
- (void) stopInAppMessageLoadTimer {
    if (nil != self.inAppMsgTimer) {
        [self.inAppMsgTimer invalidate];
        self.inAppMsgTimer = nil;
    }
}



// handle In-App msg.
- (void) handlePendingInAppMessage {
    [self fetchInAppNotificationsFromDataStore: BlueShiftInAppTriggerUpComing];
}


// trigger queued notifications
- (void)scanNotificationQueue {
    if (self.notificationControllerQueue && [self.notificationControllerQueue count] > 0) {
        BlueShiftNotificationViewController *notificationController = [self.notificationControllerQueue objectAtIndex:0];
        [self.notificationControllerQueue removeObjectAtIndex:0];
        [self presentInAppNotification:notificationController];
    }
}


// Present ViewController
- (void)presentInAppNotification:(BlueShiftNotificationViewController*)notificationController {
    if (self.currentNotificationController) {
        // if we are currently displaying a notification, queue this notification for later display
        [self.notificationControllerQueue addObject:notificationController];
        return;
    } else {
        // no current notification so display
        self.currentNotificationController = notificationController;
        [notificationController show:YES];
    }
}


- (void)createNotification:(BlueShiftInAppNotification*)notification {
    BlueShiftNotificationViewController *notificationController;
    NSString *errorString = nil;
    
    switch (notification.inAppType) {
        case BlueShiftInAppTypeHTML:
            notificationController = [[BlueShiftNotificationWebViewController alloc] initWithNotification:notification];
            break;
        case BlueShiftInAppTypeModal:
            notificationController = [[BlueShiftNotificationModalViewController alloc] initWithNotification:notification];
            break;
        case BlueShiftNotificationSlideBanner:
            notificationController = [[BlueShiftNotificationSlideBannerViewController alloc] initWithNotification:notification];
            break;
            
        default:
            errorString = [NSString stringWithFormat:@"Unhandled notification type: %lu", (unsigned long)notification.inAppType];
            break;
    }
    if (notificationController) {
        notificationController.delegate = self;
        [notificationController setTouchesPassThroughWindow: notification.templateStyle.enableBackgroundAction];
        [self presentInAppNotification:notificationController];
    }
    if (errorString) {
    }
    
}

- (void)createNotificationFromDictionary:(InAppNotificationEntity *) inAppEntity {
    
    BlueShiftInAppNotification *inAppNotification = [[BlueShiftInAppNotification alloc] initFromEntity:inAppEntity];
    
    if (inAppNotification.inAppType == BlueShiftNotificationRating) {
        
        [self DisplayReviewController];
    } else {
        [self createNotification: inAppNotification];
    }
}


- (void)DisplayReviewController {
    if (@available(iOS 10.3, *)) {
        [SKStoreReviewController requestReview];
    }
}


// Notification Click Callbacks
-(void)inAppDidDismiss:(BlueShiftInAppNotification *)notificationPayload fromViewController:(BlueShiftNotificationViewController *)controller  {
    
    NSManagedObjectID* entityItem = controller.notification.objectID;
    
    self.currentNotificationController = nil;
    
    /* delete the app entity from core data */
    [self removeInAppNotificationFromDB: entityItem];
    
    /* scan queue for any pending notification. */
    //TODO:  check app foreground state before scanning.
    [self scanNotificationQueue];
    
    //[[self inAppNotificationDelegate] dismissButtonDidTapped: notificationPayload];
}

-(void)inAppActionDidTapped:(NSDictionary *)notificationPayload fromViewController:(BlueShiftNotificationViewController *)controller {
    [[self inAppNotificationDelegate] actionButtonDidTapped: notificationPayload];
}

// Notification render Callbacks
-(void)inAppDidShow:(BlueShiftInAppNotification *)notification fromViewController:(BlueShiftNotificationViewController *)controller {
    
}

@end


BlueShift-iOS-SDK/InApps/BlueShiftInAppNotificationConstant.h
//
//  BlueShiftInAppNotificationConstant.h
//  Pods
//
//  Created by Noufal Subair on 31/07/19.
//

#ifndef BlueShiftInAppNotificationConstant_h
#define BlueShiftInAppNotificationConstant_h

/* InApp Notification Modal */
#define kInAppNotificationKey                                   @"inapp"
#define kInAppNotificationModalDismissButtonKey                 @"action.dismiss"
#define kInAppNotificationModalAppOpenButtonKey                 @"action.app_open"
#define kInAppNotificationModalShareButtonKey                   @"action.share"
#define kInAppNotificationModalContentKey                       @"content"
#define kInAppNotificationModalHTMLKey                          @"html"
#define kInAppNotificationModalURLKey                           @"url"
#define kInAppNotificationModalTitleKey                         @"title"
#define kInAppNotificationModalSubTitleKey                      @"sub_title"
#define kInAppNotificationModalMessageKey                       @"message"
#define kInAppNotificationModalBackgroundImageKey               @"background_image"
#define kInAppNotificationModalBackgroundColorKey               @"background_color"
#define kInAppNotificationModalIconKey                          @"icon"

#define kInAppNotificationModalContentStyleKey                  @"content_style"
#define kInAppNotificationModalTitleColorKey                    @"title_color"
#define kInAppNotificationModalTitleBackgroundColorKey          @"title_background_color"
#define kInAppNotificationModalTitleGravityKey                  @"title_gravity"
#define kInAppNotificationModalTitleSizeKey                     @"title_size"
#define kInAppNotificationModalMessageColorKey                  @"message_color"
#define kInAppNotificationModalMessageAlignKey                  @"message_align"
#define kInAppNotificationModalMessageBackgroundColorKey        @"message_background_color"
#define kInAppNotificationModalMessageGravityKey                @"message_gravity"
#define kInAppNotificationModalMessageSizeKey                   @"message_size"
#define kInAppNotificationModalIconSizeKey                      @"icon_size"
#define kInAppNotificationModalIconColorKey                     @"icon_color"
#define kInAppNotificationModalIconBackgroundColorKey           @"icon_background_color"
#define kInAppNotificationModalIconBackgroundRadiusKey          @"icon_background_radius"

#define kInAppNotificationModalTemplateStyleKey                 @"template_style"
#define kInAppNotificationModalPositionKey                      @"position"
#define kInAppNotificationModalWidthKey                         @"width"
#define kInAppNotificationModalHeightKey                        @"height"
#define kInAppNotificationModalFullScreenKey                    @"fullscreen"
#define kInAppNotificationModalBackgroundActionKey              @"enable_background_action"

#define kInAppNotificationModalTextKey                          @"text"
#define kInAppNotiificationModalTextColorKey                    @"text_color"
#define kInAppNotificationModalPageKey                          @"page"
#define kInAppNotificationModalExtraKey                         @"extra"
#define kInAppNotificationModalImageKey                         @"image"

/* Define a Font */
#define kInAppNotificationModalFontAwesomeFileNameKey           @"FontAwesome"
#define kInAppNotificationModalFontExtensionKey                 @"otf"
#define kInAppNotificationModalFontAwesomeNameKey               @"FontAwesome5Free-Solid"

/* Define XIB Name */
#define kInAppNotificationModalXIBNameKey                       @"BlueshiftNotificationModal"
#define kInAppNotificationModalWithImageXIBNameKey              @"BlueshiftNotificationModalWithImage"
#define kInAppNotificationSlideBannerXIBNameKey                 @"BlueShiftNotificationSlideBanner"
#define kInAppNotificationModalWithOneButtonXIBNameKey          @"BlueShiftNotificationModalWithOneButton"

/* Define Position Modal */
#define kInAppNotificationModalPositionBottomKey                @"bottom"
#define kInAppNotificationModalPositionTopKey                   @"top"
#define kInAppNotificationModalPositionCenterKey                @"center"

/* Define Resolution Type */
#define kInAppNotificationModalResolutionPointsKey             @"points"
#define kInAppNotificationModalResolutionPercntageKey           @"percentage"
#define kInAppNotificationModalHTMLHeaderKey                    @"<header><meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0, user-scalable=no'></header>"

/* Define a Notification Type */
#define kInAppNotificationTypeCenterPopUpKey                    @"center_popup"
#define kInAppNotificationTypeCenterPopUpImageKey               @"center_popup_with_image"
#define kInAppNotificationTypeSlideBannerKey                    @"slide_in_banner"
#define kInAppNotificationTypeRatingKey                         @"rating"
#define kInAppNotificationTypeCenterPopUpOneButtonKey           @"center_popup_one_button"

/*  Define a DataMart */
#define kInAppNotificationEntityNameKey                         @"InAppNotificationEntity"


#endif /* BlueShiftInAppNotificationConstant_h */
